Deprecate `ServiceDiscoveryRetryStrategy`

chemicL opened the pull request:
Motivation:

`ServiceDiscoveryRetryStrategy` has two responsibilities. One,
implied by the name is retrying errors from the `Publisher`
probided by `ServiceDiscoverer#discover` method. The other, implicit, is
deduplication logic for incoming events. As `RoundRobinLoadBalancer`
recently gained the ability to deduplicate events on its own, this
second responsibility is no longer a requirement. At the same time,
retries can be achieved by `Completable#retryWhen` mechanism using
standard utilities provided by `RetryStrategies`. Therefore, the
intention is to eventually remove `ServiceDiscoveryRetryStrategy` and
its' implementation, `DefaultServiceDiscoveryRetryStrategy`.

Modifications:

- Deprecated `ServiceDiscoveryRetryStrategy` and
  `DefaultServiceDiscoveryRetryStrategy`,
- Deprecated
  `PartitionedHttpClientBuilder#retryServiceDiscoveryErrors(ServiceDiscoveryRetryStrategy)`
  and introduced a variant accepting `BiIntFunction<Throwable, ? extends
  Completable>`,
- Deprecated
  `SingleAddressHttpClientBuilder#retryServiceDiscoveryErrors(ServiceDiscoveryRetryStrategy)`
  and introduced a variant accepting `BiIntFunction<Throwable, ? extends
  Completable>`.

Result:

Complicated and unnecessary `ServiceDiscoveryRetryStrategy` can be
removed in the future and users can rely on standard mechanisms for
retrying `ServiceDiscoverer` errors.

bondolo committed:
Use connection/context executor for timeout
Motivation:
Unless an executor is provided at the time of filter creation the
timeout filter will execute the timeout on the global executor. In many
cases it would be better to use the configured executor associated with
the connection or context.
Modifications:
Provide the connection/context executor to timeout creation to be used
if no executor is defined for filter.
Result:
More predictable execution of timeout.

bondolo committed:
Improve POM/BOM publication metadata
Motivation:
Some additional POM sections can be added and we can improve the
configurability of the `scm`, `issueManagement` and `ciManagement`
sections.
Modifications:
Add `issueManagement` and `ciManagement` sections to POM/BOM
definition. Add new `gradle.properties` definitions for the
values of the metadata.
Result:
Richer BOM/POM file metadata

madrob committed:
Replace deprecated Gradle 7 features (#1918)

Motivation:

To ensure compatibility with an eventual upgrade to Gradle 8

Modifications:

- Replace JavaExec.main with mainClass
- Replace appendix with archiveAppendix

Result:

Gradle deprecation warnings go away!

Co-authored-by: Mike Drob <mdrob@apple.com>

Commit:
Deprecate `ServiceDiscoveryRetryStrategy`

Motivation:

`ServiceDiscoveryRetryStrategy` has two responsibilities. One,
implied by the name is retrying errors from the `Publisher`
probided by `ServiceDiscoverer#discover` method. The other, implicit, is
deduplication logic for incoming events. As `RoundRobinLoadBalancer`
recently gained the ability to deduplicate events on its own, this
second responsibility is no longer a requirement. At the same time,
retries can be achieved by `Completable#retryWhen` mechanism using
standard utilities provided by `RetryStrategies`. Therefore, the
intention is to eventually remove `ServiceDiscoveryRetryStrategy` and
its' implementation, `DefaultServiceDiscoveryRetryStrategy`.

Modifications:

- Deprecated `ServiceDiscoveryRetryStrategy` and
  `DefaultServiceDiscoveryRetryStrategy`,
- Deprecated
  `PartitionedHttpClientBuilder#retryServiceDiscoveryErrors(ServiceDiscoveryRetryStrategy)`
  and introduced a variant accepting `BiIntFunction<Throwable, ? extends
  Completable>`,
- Deprecated
  `SingleAddressHttpClientBuilder#retryServiceDiscoveryErrors(ServiceDiscoveryRetryStrategy)`
  and introduced a variant accepting `BiIntFunction<Throwable, ? extends
  Completable>`.

Result:

Complicated and unnecessary `ServiceDiscoveryRetryStrategy` can be
removed in the future and users can rely on standard mechanisms for
retrying `ServiceDiscoverer` errors.

Commit:
Simplified default strategy selection logic

tkountis reviewed (approved):
LGTM

idelpivnitskiy committed:
Add missed `serialVersionUID` for exceptions (#1919)

Motivation:

`Serializable` interface requires `serialVersionUID`.

Modifications:

- Add `serialVersionUID` for `ConnectionClosedException`,
`ProxyResponseException`, `StacklessClosedChannelException`;

Result:

Exceptions are aligned with `Serializable` contract.

bondolo committed:
improve variable names and add documentation

bondolo committed:
Add `ExecutionStrategyInfluencer` (#1832)

Motivation:
To allow for simpler determination of the required execution strategy
for collections of client and server code.
Modifications:
Introduce `ExecutionStrategyInfluencer` and `requiredOffloads()` as
a replacement for the `influenceStrategy` method of
`HttpExecutionStrategyInfluencer` which is also deprecated.
Result:
A path towards simplified and more transparent execution
strategy computations.

idelpivnitskiy commented on servicetalk-http-api/src/main/java/io/servicetalk/http/api/ServiceDiscoveryRetryStrategy.java:
```diff
@@ -26,7 +27,13 @@
+ * @see SingleAddressHttpClientBuilder#retryServiceDiscoveryErrors(BiIntFunction)
+ * @deprecated Use standard mechanisms for retrying using
+ * {@link io.servicetalk.concurrent.api.Completable#retryWhen(BiIntFunction)} on the {@link Publisher} returned from
+ * {@link ServiceDiscoverer#discover(Object)} instead.
```
This is a misleading suggestion. Users should not retry on `ServiceDiscoverer#discover(Object)` because it will prevent client from seeing SD errors. Consider pointing to the new builder overloads that take `BiIntFunction`

idelpivnitskiy commented on servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java:
```diff
@@ -52,7 +52,13 @@
+ * @see io.servicetalk.concurrent.api.RetryStrategies
+ * @see SingleAddressHttpClientBuilder#retryServiceDiscoveryErrors(BiIntFunction)
+ * @deprecated Use standard mechanisms for retrying using
+ * {@link io.servicetalk.concurrent.api.Completable#retryWhen(BiIntFunction)} on the {@link Publisher} returned from
```
Same

idelpivnitskiy commented on servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientBuilder.java:
```diff
@@ -97,9 +99,25 @@
      */
     PartitionedHttpClientBuilder<U, R> retryServiceDiscoveryErrors(
-            ServiceDiscoveryRetryStrategy<R, PartitionedServiceDiscovererEvent<R>> retryStrategy);
+            BiIntFunction<Throwable, ? extends Completable> retryStrategy);
```
When you cherry-pick, make sure a new method has a default impl in 0.41 branch

idelpivnitskiy commented on servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:
```diff
@@ -291,9 +293,25 @@
      */
     SingleAddressHttpClientBuilder<U, R> retryServiceDiscoveryErrors(
-            ServiceDiscoveryRetryStrategy<R, ServiceDiscovererEvent<R>> retryStrategy);
+            BiIntFunction<Throwable, ? extends Completable> retryStrategy);
```
When you cherry-pick, make sure a new method has a default impl in 0.41 branch

Scottmitch committed:
Enhance javadoc to clarify StackOverflowError possibility (#1925)

Motivation:
Publisher [repeat | retry] operators may cause StackOverflowError if too
many recovery operations are done consecutively on the same thread. This
condition can be difficult to detect as the onError recovery path is
unlikely to be called. We should highlight this possibility in the
javadocs for these methods.

Modifications:
- Publisher, Single, Completable repeat* and retry* operators have
  javadoc that clarifies the risk of StackOverflowError and mitigation
  technique.

Result:
Javadocs highlight risks of StackOverflowError and how it can be
avoided.

idelpivnitskiy commented on servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:
```diff
@@ -538,6 +555,13 @@ private static StreamingHttpConnectionFilterFactory appendConnectionFilter(
     @Override
     public DefaultSingleAddressHttpClientBuilder<U, R> retryServiceDiscoveryErrors(
             final ServiceDiscoveryRetryStrategy<R, ServiceDiscovererEvent<R>> retryStrategy) {
+        this.deprecatedServiceDiscovererRetryStrategy = requireNonNull(retryStrategy);
```
Instead of opening a possibility for users to have two strategies at the same time, consider nullifying the other one when a builder method is invoked. Clarify that in javadoc. You can also log at warn level if you feel it's necessary.

idelpivnitskiy commented on servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultPartitionedHttpClientBuilder.java:
```diff
@@ -279,7 +290,14 @@ public StreamingHttpRequest newRequest(final HttpRequestMethod method, final Str
+    @Override
+    public PartitionedHttpClientBuilder<U, R> retryServiceDiscoveryErrors(
+            final BiIntFunction<Throwable, ? extends Completable> retryStrategy) {
+        this.serviceDiscovererRetryStrategy = retryStrategy;
```
Same, also use `requireNonNull`

cancecen committed:
Enhance `GrpcStatus.fromCodeValue` static factories to include actual status code when it maps to `UNKNOWN` (#1923)

Motivation:
Preserving the original code value in the description when the code is not recognized. This is also consistent with grpc-java.

Modifications:
Pass in the original error code to `GrpcStatus` in the static factories.

Result:
`GrpcStatus` for unrecognized status codes preserves the ogirinal code value.

vglushak committed:
Remove deprecated method from `AwaitUtils` (#1927)

Motivation: code cleanup.
Modifications: removed unused deprecated methods.
Result: AwaitUtils.java 225 lines -> 107 lines.

Scottmitch committed:
MalformedDataAfterHttpMessageTest.afterResponseNextClientRequestSucceeds fix (#1926)

Motivation:
afterResponseNextClientRequestSucceeds attempts to validate that
NettyChannelPublisher preemptively closes the channel after request
processing concluded if there is an error pending to prevent subsequent
requests from using this connection. This condition may not be hit in
all scenarios due to sequencing of offloading as well as when bytes are
read from the network and the error condition is observed. This results
in the test being flaky and failing.

Modifications:
- Make afterResponseNextClientRequestSucceeds more robust by inserting a
  RetryingHttpRequesterFilter and adjusting the assertions to expect
  retry may happen.

Result:
Fixes https://github.com/apple/servicetalk/issues/1889.

Scottmitch committed:
Make static final buffers unreleasable and read-only (#1928)

Motivation:
`HttpObjectEncoder` writes static final ByteBufs to encode HTTP chunks.
Some handlers (e.g. SslHandler) may attempt to aggregate data into
existing buffers if they can be expanded. If static buffers are not
marked as ReadOnly then data maybe appended to these static buffers over
time which may lead to data corruption.

Modification:
- Make static ByteBuf/Buffer objects ReadOnly in `HttpObjectEncoder` and
  elsewhere.

Result:
Static buffers will not be used to aggregate data.

idelpivnitskiy committed:
Deprecate unused exception types (#1920)

Motivation:

There are a few custom exception types that were used in the past but
became unusable after some refactoring. It's safe to deprecate them and
remove in future releases.

Modifications:

- Deprecate `ConnectionClosedException`, `InvalidRedirectException`
`MaxRequestLimitExceededException`,
`MaxRequestLimitExceededRejectedSubscribeException`;
- Avoid using deprecated exceptions in tests;

Result:

All unused exception types are deprecated now.

idelpivnitskiy committed:
Fixes for `EmptyBuffer` implementation (#1929)

Motivation:

`EmptyBuffer` has a few minor issues.

Modifications:

- Return a read-only view of `ByteBuffer` from `toNioBuffer()`;
- Implement `toString()`, `equals(Object)`, and `hashCode()`;

Result:

`EmptyBuffer` correctly implements methods from `Object` and is
consistent with netty's `EmptyByteBuf` implementation.

bondolo committed:
Remove `grpc` and `http` modules from BOM (#1930)

Motivation:
The ServiceTalk BOM file currently contains references to two modules,
`grpc` and `http` which do not exist. These modules are generated from
the examples parent directories but have no artifacts.
Modifications:
Exclude the `grpc` and `http` examples parent modules when building
BOM.
Result:
Non-existant modules no longer included in BOM.

idelpivnitskiy committed:
Remove unused exception types (#1922)

Motivation:

There are a few custom exception types that are unused and were
deprecated in #1920. We can remove them now.

Modifications:

- Remove `ConnectionClosedException`, `InvalidRedirectException`
`MaxRequestLimitExceededException`,
`MaxRequestLimitExceededRejectedSubscribeException`;

Result:

No unused exception types.

Commit:
Merge branch 'main' into deprecate-service-discovery-retry-strategy

Commit:
Added default implementations for new methods, improved javadocs

chemicL commented:
Thanks @idelpivnitskiy and @tkountis. I changed the base branch to main and will open this PR in the source repo.
